Add `allowChildren` schema item definition property

jacekbogdanski opened the pull request:
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Features (engine): Introduced the `SchemaItemDefinition#allowChildren` property simplifying defining which other items are allowed inside this schema item definition. Closes #9261.

---

### Additional information

#### Missing definition

Some time ago we've discussed that in case of missing parent schema item definition, this feature should throw an error. However, looking at how we are currently handling such cases, it seems that it should rather ensure that the feature works correctly even if the item defined in `allowChildren` does not exist. As an example, take a look at `allowIn` unit tests:

https://github.com/ckeditor/ckeditor5/blob/c9ddd79b441fb358d7c1a65b1aa44a71dd29cafd/packages/ckeditor5-engine/tests/model/schema.js#L2744-L2751

Thanks to that we can still register definitions lazily without specific order:


```js
// We are registering paragraph before $root and without $text but this code will still work.
schema.register( 'paragraph', {
  isBlock: true,
  allowChildren: '$text'
} );

schema.register( '$root', {
  allowChildren: 'paragraph'
} );
```

We could still keep this behavior (lazy loading) by throwing errors during compilation, not definition registration. However, it will work differently than the current approach for other properties, so it probably makes more sense to avoid throwing errors in case of missing definitions and just skip it.

#### Refactoring

In some places, we are using the approach of extending the existing definition (mostly for `$text` item). As an example:

https://github.com/ckeditor/ckeditor5/blob/c9ddd79b441fb358d7c1a65b1aa44a71dd29cafd/packages/ckeditor5-engine/src/model/model.js#L107-L111

I'm wondering if it makes sense to refactor such places, as it could be rewritten to:

```js
this.schema.register( '$clipboardHolder', {
	allowContentOf: '$root',
	allowChildren: '$text',
	isLimit: true
} );
```
I didn't want to do it before the first review iteration to avoid unnecessary changes for now, but I suppose it would be nice to improve the code if we already have a bit nicer API for that. WDYT?

niegowski commented on packages/ckeditor5-engine/tests/model/schema.js:
```diff
@@ -2411,6 +2411,173 @@ describe( 'Schema', () => {
+				expect( schema.checkChild( div, paragraph ) ).to.be.false;
+			} );
+
+			it( 'should remove allowChildren from definition', () => {
+				schema.register( '$root', {
+					allowChildren: 'paragraph'
+				} );
+
+				schema.register( 'paragraph' );
+
+				expect( schema.getDefinition( '$root' ) ).to.deep.equal( {
+					allowAttributes: [],
+					allowIn: [],
+					name: '$root'
+				} );
+			} );
```
Maybe it could be useful to be able to see what elements might be children for some element?

jacekbogdanski replied:
Seeing all allowed children for the specific definition looks very useful for _at least_ debugging, even if we don't have any real case scenario for now 👍🏻

niegowski reviewed (commented):
As commented below. I would also update the usages in our features. The guides should also be updated.

jacekbogdanski commented:
I've added `allowChildren` to a compiled schema item definition, refactored code to use new property where it makes sense (didn't cover every possible unit test as in some cases it's more readable when actually the code is using a longer `extend` approach) and updated guide.

godai78 commented on packages/ckeditor5-engine/docs/framework/guides/deep-dive/schema.md:
```diff
@@ -17,7 +17,7 @@ Elements and attributes are checked by features separately by using the {@link m
 ## Defining allowed structures
 
-When a feature introduces a model element, it should register it in the schema. Besides defining that such an element may exist in the model, the feature also needs to define where this element can be placed:
+When a feature introduces a model element, it should register it in the schema. Besides defining that such an element may exist in the model, the feature also needs to define where this element can be placed. This information is provided by {@link module:engine/model/schema~SchemaItemDefinition#allowIn} property of the {@link module:engine/model/schema~SchemaItemDefinition}:
```
```suggestion
When a feature introduces a model element, it should register it in the schema. Besides defining that such an element may exist in the model, the feature also needs to define where this element can be placed. This information is provided by the {@link module:engine/model/schema~SchemaItemDefinition#allowIn} property of the {@link module:engine/model/schema~SchemaItemDefinition}:
```

godai78 commented on packages/ckeditor5-engine/docs/framework/guides/deep-dive/schema.md:
````diff
@@ -37,14 +37,39 @@ In other words, this would be correct:
 </$root>
 ```
 
+To declare what nodes could be inside the registered element, the {@link module:engine/model/schema~SchemaItemDefinition#allowChildren} property could be used:
````
```suggestion
To declare which nodes could be inside the registered element, the {@link module:engine/model/schema~SchemaItemDefinition#allowChildren} property could be used:
```

godai78 commented on packages/ckeditor5-engine/docs/framework/guides/deep-dive/schema.md:
````diff
@@ -37,14 +37,39 @@ In other words, this would be correct:
+</$root>
+```
+
+Both `{@link module:engine/model/schema~SchemaItemDefinition#allowIn}` and `{@link module:engine/model/schema~SchemaItemDefinition#allowChildren}` properties can be also inherited from other `SchemaItemDefinition` items. 
````
```suggestion
Both the `{@link module:engine/model/schema~SchemaItemDefinition#allowIn}` and `{@link module:engine/model/schema~SchemaItemDefinition#allowChildren}` properties can be also inherited from other `SchemaItemDefinition` items. 
```

jacekbogdanski commented on packages/ckeditor5-engine/src/model/schema.js:
```diff
@@ -1654,6 +1693,16 @@ function cleanUpAllowIn( compiledDefinitions, itemName ) {
 }
 
+// Restore allowChildren items based on allowIn.
+function cleanUpAllowChildren( compiledDefinitions, itemName ) {
```
I was wondering about `setupAllowChildren` and `restoreAllowChildren` - I like `restore` a bit less but `setup` would be too close to `compile` and hard to distinguish two methods.

niegowski reviewed (approved):
👍  LGTM

godai78 reviewed (approved):
godai's seal of approval
